enhancements/management-workload-partitioning.md

```diff
@@ -0,0 +1,848 @@
+2. The user creates a machine config manifest to configure CRI-O to
+   partition management workloads.
+3. The user creates a machine config manifest to write a configuration
+   file for kubelet to enable the same workload partition. The file
+   should only be readable by the kubelet.
+4. The user runs the installer to create the standard manifests, adds
+   their extra manifests from steps 2 and 3, then creates the cluster.
+5. The kubelet starts up and finds the configuration file enabling the
+   new feature.
+6. The kubelet advertises `management.workload.openshift.io/cores`
+   extended resources on the node based on the number of CPUs in the
+   host.
+7. The kubelet reads static pod definitions. It replaces the `cpu`
+   requests with `cpu.workload.openshift.io/management` requests of
+   the same value and adds the
+   `io.openshift.workload.management.cpushares/{container-name}`
+   annotations for CRI-O with the same values.
+8. Something schedules a regular pod with the
+   `workload.openshift.io/management` annotation in a namespace with
+   the `workload.openshift.io/allowed=management` annotation.
+9. The admission hook modifies the pod, replacing the CPU requests
+   with `management.workload.openshift.io/cores` requests and adding
+   the `io.openshift.workload.management.cpushares/{container-name}`
+   annotations for CRI-O.
+10. The scheduler sees the new pod and finds available
+    `management.workload.openshift.io/cores` resources on the node. The
+    scheduler places the pod on the node.
+11. Repeat steps 8-10 until all pods are running.
+12. Single-node deployment comes up with management components
+    constrained to subset of available CPUs.
+
+#### Workload Annotation
+
+The `workload.openshift.io` annotation on each pod needs to allow us
+to add new parameters in the future, so the value will be a
+struct. Initially, it will encode 2 values.
+
+The *workload type* for the workloads is the suffix of the annotation
+name to make it easy for CRI-O and other lower-level components to
+detect it without understanding the full JSON struct in the value.
+
+The `effect` field will control whether the request is a soft or hard
+rule. It can contain either `PreferredDuringScheduling` for soft
+requests or `RequiredDuringScheduling` for hard requests. Initially we
+will only implement the soft request mode and that will be the
+default.
+
+```yaml
+metadata:
+  annotations:
+    workload.openshift.io/management: |
+      {"effect": "PreferredDuringScheduling"}
+```
+
+The admission hook will reject pods with multiple annotations, for
+now. Future work may allow multiple workload types with different
+priorities to support clusters with different types of configurations.
+
+#### Pod mutation
+
+The kubelet and API admission hook will change pods annotated with
+`workload.openshift.io/management` so the CPU requests are replaced
+with management CPU requests and an annotation is added with the same
```

Comment:
The cluster-version operator's reconciliation currently [requires in-cluster `requests` to `DeepEqual` the manifest's requests][1], so it's going to be hotlooping on this trying to stomp out the admission hook's change.  We have one CVO-hotloop bug where [one potential approach is server-side apply][2], which should work around that (unless I'm misunderstanding server-side apply).  But looks like server-side apply [is still beta][3].  If we need this to work before we have stable server-side apply, do folks have recommendations for updating the CVO's `requests`-reconciliation logic?  Do we need to teach it about what the admission hook might do, so it can ignore that?  Something less invasive?  I expect this is going to be an issue for every operator which manages operand workloads, not just for the CVO.

[1]: https://github.com/openshift/cluster-version-operator/blob/master/lib/resourcemerge/core.go#L522-L530
[2]: https://bugzilla.redhat.com/show_bug.cgi?id=1881484#c1
[3]: https://kubernetes.io/docs/reference/using-api/server-side-apply/

Reply:
@dhellmann pointed out that the admission hook will only mutate pods, not Deployments, etc., so the CVO won't notice.  I dunno how the Deployment and DaemonSet controllers handle reconciling pod requests, but that's not my code to worry about ;)

Reply:
According to @derekwaynecarr those controllers allow mutations like this. That's how service mesh layers work, for example.